src/platform_impl/linux/x11/ime/context.rs

```diff
@@ -1,41 +1,239 @@
+            CompositionUpdate(
+                String::from_iter(client_data.text.clone()),
+                client_data.cursor_pos,
+            ),
+        ))
+        .expect("failed to send composition update event");
+}
+
+extern "C" fn preedit_caret_callback(
+    _xim: ffi::XIM,
+    client_data: ffi::XPointer,
+    call_data: ffi::XPointer,
+) {
+    let client_data = unsafe { &mut *(client_data as *mut ImeContextClientData) };
+    let call_data = unsafe { &mut *(call_data as *mut XIMPreeditCaretCallbackStruct) };
+    client_data.cursor_pos = call_data.position as usize;
+
+    client_data
+        .event_sender
+        .send((
+            client_data.window,
+            CompositionUpdate(
+                String::from_iter(client_data.text.clone()),
+                client_data.cursor_pos,
+            ),
+        ))
+        .expect("failed to send composition update event");
+}
 
 unsafe fn create_pre_edit_attr<'a>(
+    xconn: &'a Arc<XConnection>,
+    preedit_callbacks: &'a PreeditCallbacks,
+) -> util::XSmartPointer<'a, c_void> {
+    util::XSmartPointer::new(
+        xconn,
+        (xconn.xlib.XVaCreateNestedList)(
+            0,
+            ffi::XNPreeditStartCallback_0.as_ptr() as *const _,
+            &(preedit_callbacks.start_callback) as *const _,
+            ffi::XNPreeditDoneCallback_0.as_ptr() as *const _,
+            &(preedit_callbacks.done_callback) as *const _,
+            ffi::XNPreeditCaretCallback_0.as_ptr() as *const _,
+            &(preedit_callbacks.caret_callback) as *const _,
+            ffi::XNPreeditDrawCallback_0.as_ptr() as *const _,
+            &(preedit_callbacks.draw_callback) as *const _,
+            ptr::null_mut::<()>(),
+        ),
+    )
+    .expect("XVaCreateNestedList returned NULL")
+}
+
+unsafe fn create_pre_edit_attr_with_spot<'a>(
     xconn: &'a Arc<XConnection>,
     ic_spot: &'a ffi::XPoint,
+    preedit_callbacks: &'a PreeditCallbacks,
 ) -> util::XSmartPointer<'a, c_void> {
     util::XSmartPointer::new(
         xconn,
         (xconn.xlib.XVaCreateNestedList)(
             0,
             ffi::XNSpotLocation_0.as_ptr() as *const _,
             ic_spot,
+            ffi::XNPreeditStartCallback_0.as_ptr() as *const _,
+            &preedit_callbacks.start_callback as *const _,
+            ffi::XNPreeditDoneCallback_0.as_ptr() as *const _,
+            &preedit_callbacks.done_callback as *const _,
+            ffi::XNPreeditCaretCallback_0.as_ptr() as *const _,
+            &preedit_callbacks.caret_callback as *const _,
+            ffi::XNPreeditDrawCallback_0.as_ptr() as *const _,
+            &preedit_callbacks.draw_callback as *const _,
             ptr::null_mut::<()>(),
         ),
     )
     .expect("XVaCreateNestedList returned NULL")
 }
 
+fn create_xim_callback(client_data: ffi::XPointer, callback: XIMProcNonnull) -> ffi::XIMCallback {
+    let mut xim_callback: ffi::XIMCallback = unsafe { MaybeUninit::uninit().assume_init() };
```

Comment:
This is not a safe usage of `MaybeUninit`. Also, it appears to be totally unnecessary. The `XIMCallback` struct instance can be constructed in the usual manner.

Reply:
Fixed